tests/cmds/test_auditlogs.py

```diff
@@ -92,15 +199,206 @@ def test_search_audit_logs_with_all_filter_parameters(runner, cli_state, date_st
+    )
+    assert (
+        cli_state.sdk.auditlogs.get_all.call_args[1]["begin_time"] == CURSOR_TIMESTAMP
+    )
+
+
+def test_send_to_with_existing_checkpoint_replaces_begin_arg_if_passed(
+    cli_state, runner, test_audit_log_response, audit_log_cursor_with_checkpoint
+):
+    runner.invoke(
+        cli,
+        [
+            "audit-logs",
+            "send-to",
+            "localhost",
+            "--begin",
+            "1d",
+            "--use-checkpoint",
+            "test",
+        ],
+        obj=cli_state,
+    )
+    assert (
+        cli_state.sdk.auditlogs.get_all.call_args[1]["begin_time"] == CURSOR_TIMESTAMP
+    )
+
+
+def test_search_with_existing_checkpoint_events_skips_duplicate_events(
+    cli_state,
+    runner,
+    test_audit_log_response,
+    audit_log_cursor_with_checkpoint_and_events,
+):
+    cli_state.sdk.auditlogs.get_all.return_value = [test_audit_log_response]
+    result = runner.invoke(
+        cli,
+        ["audit-logs", "search", "--begin", "1d", "--use-checkpoint", "test"],
+        obj=cli_state,
+    )
+    assert "42@code42.com" not in result.stdout
+    assert "43@code42.com" in result.stdout
+
+
+def test_send_to_with_existing_checkpoint_events_skips_duplicate_events(
+    cli_state,
+    runner,
+    test_audit_log_response,
+    audit_log_cursor_with_checkpoint_and_events,
+    send_to_logger,
+):
+    cli_state.sdk.auditlogs.get_all.return_value = [test_audit_log_response]
+    runner.invoke(
+        cli,
+        [
+            "audit-logs",
+            "send-to",
+            "localhost",
+            "--begin",
+            "1d",
+            "--use-checkpoint",
+            "test",
+        ],
+        obj=cli_state,
+    )
+    assert send_to_logger.info.call_count == 3
+    assert send_to_logger.info.call_args_list[0][0][0]["actorName"] != "42@code42.com"
+
+
+def test_search_without_existing_checkpoint_writes_both_event_hashes_with_same_timestamp(
+    cli_state,
+    runner,
+    test_audit_log_response_with_only_same_timestamps,
+    audit_log_cursor_with_checkpoint,
+):
+    cli_state.sdk.auditlogs.get_all.return_value = [
+        test_audit_log_response_with_only_same_timestamps
+    ]
+    runner.invoke(
+        cli,
+        ["audit-logs", "search", "--begin", "1d", "--use-checkpoint", "test"],
+        obj=cli_state,
+    )
+    assert audit_log_cursor_with_checkpoint.replace_events.call_count == 2
+    assert audit_log_cursor_with_checkpoint.replace_events.call_args_list[1][0][1] == [
+        hash_event(TEST_EVENTS_WITH_SAME_TIMESTAMP[0]),
+        hash_event(TEST_EVENTS_WITH_SAME_TIMESTAMP[1]),
+    ]
+
+
+def test_send_to_without_existing_checkpoint_writes_both_event_hashes_with_same_timestamp(
+    cli_state,
+    runner,
+    test_audit_log_response_with_only_same_timestamps,
+    audit_log_cursor_with_checkpoint,
+    send_to_logger,
+):
+    cli_state.sdk.auditlogs.get_all.return_value = [
+        test_audit_log_response_with_only_same_timestamps
```

Comment:
Two things about these tests make me nervous:

1.) The mock get_all responses should actually be generator instead of lists

2.)  I don't see any tests that make use of more than 1 page.

Reply:
Updated to use generator and to have tests that use more than one page.